File: adventofcode/14/14.cpp
#include <bits/stdc++.h>
using namespace std;

int main() {
    int n;
    cin >> n;
    map<long long, long long> m1, m2;
    long long zeros, ones, xs, sol1 = 0, sol2 = 0;
    vector<long long> offsets, offsets_aux; 
    for(int i = 0; i<n; i++){
        string instruction, value;
        cin >> instruction >> value >> value;
        if (instruction == "mask") {
            zeros = xs = (1L << 35)-1; ones = 0;
            offsets.clear(); offsets.push_back(0);
            for(int c = 0; c < value.size(); c++) {
                if (value[c] == '0') zeros = zeros ^ (1L << (35-c));
                if (value[c] == '1') ones = ones | (1L << (35-c));
                if (value[c] == 'X') {
                    xs = xs ^ (1L << (35-c));
                    offsets_aux = offsets;
                    for (int o = 0; o < offsets.size(); o++) offsets_aux.push_back(offsets[o] + (1L << (35-c)));
                    offsets = offsets_aux;
                }
            }
        } else {
            int pos = stoi(instruction.substr(4, instruction.size()-1));
            m1[pos] = stoi(value) & zeros | ones;
            for (int o = 0; o < offsets.size(); o++) { 
                m2[pos & xs | ones + offsets[o]] = stoi(value);
            }
        }
    }
    for (map<long long, long long>::iterator i = m1.begin(); i != m1.end(); i++) sol1 += i->second;
    for (map<long long, long long>::iterator i = m2.begin(); i != m2.end(); i++) sol2 += i->second;
    cout << "Problem1: " << sol1 << endl << "Problem2: " << sol2 << endl;
}

File: adventofcode/14/14.py
m1 = {}
m2 = {}
for line in [input().rstrip() for _ in range(int(input()))]:
    if line.split(' ')[0] == 'mask':
        mask = line.split(' ')[2]
        zeros, ones, xs = int(mask.replace('X', '1'), 2), int(mask.replace('X', '0'), 2), int(mask.replace('0', '1').replace('X', '0'), 2)
        offsets = [0]
        for i in range(0, 36):
            if mask[i] == 'X':
                for offset in offsets.copy():
                    offsets.append(offset + (1<<(35-i)))
    else:
        pos, value = int(line.split(' ')[0][4:-1]), int(line.split(' ')[2])
        m1[pos] = value & zeros | ones
        for offset in offsets:
            m2[(pos & xs | ones) + offset] = value
print("Problem1: {}\nProblem2: {}".format(sum([m1[i] for i in m1]), sum([m2[i] for i in m2])))